jovo-platforms/jovo-platform-facebookmessenger/src/FacebookMessenger.ts

```diff
@@ -290,6 +330,13 @@ export class FacebookMessenger extends Platform<MessengerBotRequest, MessengerBo
     }
   }
 
+  async afterResponse(handleRequest: HandleRequest) {
+    if (!handleRequest.jovo || handleRequest.jovo.constructor.name !== this.getAppType()) {
+      return Promise.resolve();
+    }
+    await handleRequest.host.setResponse(handleRequest.jovo.$response);
```

Comment:
I haven't tested this piece yet, but doesn't [this part](https://github.com/jovotech/jovo-framework/blob/7c5ca44703e483c45f40ced5dfc2f85cf5e43544/jovo-platforms/jovo-platform-facebookmessenger/src/FacebookMessenger.ts#L357) cause the response to always be an empty object?

Reply:
Hey @m-ripper. Nice catch! That would definitely return an empty object for request body that takes the following format

```
{
  "object":"page",
  "entry":[
    {
      "id":"<PAGE_ID>",
      "time":1458692752478,
      "messaging":[
        {
          "sender":{
            "id":"<PSID>"
          },
          "recipient":{
            "id":"<PAGE_ID>"
          },

          ...
        }
      ]
    }
  ]
}
```

What would you suggest we do? Would an option be to remove that block of code completely and just return at the end of `overrideAppHandle()`?

Reply:
The biggest issue here is, that theoretically, a single request from Facebook can contain multiple events. 

What we could do though is keep the `afterResponse()`-hook in order to set the response, but change the code in `overrideAppHandle()`.
 
 I would recommend doing the following:
 ```
     // ... other code above
      if (isMessengerRequest) {
        const promises: Array<Promise<any>> = [];
        const responses: MessengerBotResponse[] = [];
        request.entry.forEach((entry: MessengerBotEntry) => {
          const hostCopy: Host = Object.create(host.constructor.prototype);
          // tslint:disable-next-line
          hostCopy.setResponse = async function (obj: any) {
            return responses.push(obj);
          };
          for (const key in host) {
            if (host.hasOwnProperty(key)) {
              const value = (host as any)[key];
              if (key === '$request') {
                (hostCopy as any)[key] = entry;
              } else {
                (hostCopy as any)[key] =
                  typeof value === 'object' && value.constructor.name === 'Object'
                    ? { ...value }
                    : value;
              }
            }
          }
          promises.push(PROTOTYPE_BACKUP.call(this, hostCopy));
        });

        await Promise.all(promises);
        return host.setResponse(responses);
      }
      // ... other code beneath
 ```
 
I haven't tested this, but this should work.
 
The response will then always be an array though, but I think that makes sense because we can also receive multiple events.

If you have any more questions or feedback, just let me know.

Reply:
Hey @m-ripper - Thanks for the suggestion! I see the issue you're pointing out about multiple events. I've tweaked the snippet of code you suggested and returned the same structure that I suggested in the PR. So the structure for when someone sends a request with multiple events would look something like this:

```
{
    "message": [
        [
            {
                "recipient": {
                    "id": "124"
                },
                "message": {
                    "text": "hello 1"
                },
                "message_type": "RESPONSE"
            }
        ],
        [
            {
                "recipient": {
                    "id": "124"
                },
                "message": {
                    "text": "hello 2"
                },
                "message_type": "RESPONSE"
            }
        ]
    ]
}
```

Let me know what you think about this

Reply:
First of all, thanks for the adjustments.

The problem I see with this structure is just that it's not really possible to tell which messages in `message` were created based on which event. 

Let's say two events were coming in, A and B, then it's not really easy to tell which of the messages belong to A and which belong to B, which might make testing harder. 

The implementation that I suggested solves this problem because response[0] will always contain the messages for event A and response[1] will always contain the messages for event B. 

For the example above, the response would then look like this:
```
[
  {
    "message": [
      {
        "recipient": {
          "id": "124"
        },
        "message": {
          "text": "hello 1"
        },
        "message_type": "RESPONSE"
      }
    ]
  },
  {
    "message": [
      {
        "recipient": {
          "id": "124"
        },
        "message": {
          "text": "hello 2"
        },
        "message_type": "RESPONSE"
      }
    ]
  }
]
```

But I am open to discuss this. Why would you prefer a single object? 

And another small thing I thought about is maybe renaming the `message` property in the response to `messages` because it's clearer that it is an array that way.

Reply:
Thanks for the explanation! I completely agree with what you are suggesting. I'll update the PR accordingly

Reply:
Thank you very much!